// react-rdux/webpack.build.config.js
const webpack = require('webpack');

const reactExternal = {
  root: 'React',
  commonjs2: 'react',
  commonjs: 'react',
  amd: 'react'
};

const rduxExternal = {
  root: 'rdux',
  commonjs2: 'rdux',
  commonjs: 'rdux',
  amd: 'rdux'
};

const propTypesExternal = {
  root: 'prop-types',
  commonjs2: 'prop-types',
  commonjs: 'prop-types',
  amd: 'prop-types'
};

module.exports = {
  externals: {
    'react': reactExternal,
    'rdux': rduxExternal,
    'prop-types': propTypesExternal,
  },
  entry: [
    './src/react-rdux/react-rdux.js'
  ],
  output: {
    path: __dirname + "/dist/",
    filename: "react-rdux.js",
    library: 'react-rdux',
    libraryTarget: 'umd',
    umdNamedDefine: true,
  },
  module: {
    loaders: [
      {
        test: /\.(js|jsx)$/,
        exclude: /node_modules/,
        loader: 'babel-loader',
        query: {
          presets:["es2015", "react", "stage-0"]
        }
      },
    ],
  },
  plugins: [
    new webpack.optimize.OccurrenceOrderPlugin(),
    new webpack.optimize.UglifyJsPlugin({
      sourceMap: false,
      comments: false,
    }),
    new webpack.NoEmitOnErrorsPlugin(),
  ],
  resolve: {
    extensions: ['.js', '.jsx'],
  },
  devtool: false,
};
